operator/api/v1alpha1/runtimetask_types.go

```diff
@@ -0,0 +1,150 @@
+	operatorerrors "k8s.io/kubeadm/operator/errors"
+	"k8s.io/utils/pointer"
+)
+
+// RuntimeTaskSpec defines the desired state of RuntimeTask
+type RuntimeTaskSpec struct {
+	// NodeName is a request to schedule this RuntimeTask onto a specific node.
+	// +optional
+	NodeName string `json:"nodeName,omitempty"`
+
+	// RecoveryMode sets the strategy to use when a command is failed.
+	// +optional
+	RecoveryMode string `json:"recoveryMode,omitempty"`
+
+	// Commands provide the list of commands to be performed when executing a RuntimeTask on a node
+	Commands []CommandDescriptor `json:"commands"`
+}
+
+// RuntimeTaskStatus defines the observed state of RuntimeTask
+type RuntimeTaskStatus struct {
+
+	// StartTime represents time when the RuntimeTask execution was started by the controller.
+	// It is represented in RFC3339 form and is in UTC.
+	// +optional
+	StartTime *metav1.Time `json:"startTime,omitempty"`
+
+	// CurrentCommand
+	// +optional
+	CurrentCommand int32 `json:"currentCommand,omitempty"`
+
+	// CommandProgress
+	// Please note that this field is only for allowing a better visal representation of status
+	// +optional
+	CommandProgress string `json:"commandProgress,omitempty"`
+
+	// Paused indicates that the RuntimeTask is paused.
+	// +optional
+	Paused bool `json:"paused,omitempty"`
+
+	// CompletionTime represents time when the RuntimeTask was completed.
+	// It is represented in RFC3339 form and is in UTC.
+	// +optional
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+
+	// Phase represents the current phase of RuntimeTask actuation.
+	// E.g. Pending, Running, Completed, Failed etc.
+	// +optional
+	Phase string `json:"phase,omitempty"`
+
+	// ErrorReason will be set in the event that there is a problem in executing
+	// the RuntimeTasks and will contain a succinct value suitable
+	// for machine interpretation.
+	// +optional
+	ErrorReason *operatorerrors.RuntimeTaskStatusError `json:"errorReason,omitempty"`
+
+	// ErrorMessage will be set in the event that there is a problem in executing
+	// the RuntimeTasks and will contain a more verbose string suitable
+	// for logging and human consumption.
+	// +optional
+	ErrorMessage *string `json:"errorMessage,omitempty"`
+}
+
+// SetStartTime is a utility method for setting the StartTime field to Now.
+func (s *RuntimeTaskStatus) SetStartTime() {
+	now := metav1.Now()
+	s.StartTime = &now
+}
+
+// NextCurrentCommand is a utility method for setting the CurrentCommand and CommandProgress.
+func (s *RuntimeTaskStatus) NextCurrentCommand(commands []CommandDescriptor) {
+	s.CurrentCommand = s.CurrentCommand + 1
+	s.CommandProgress = fmt.Sprintf("%d/%d", s.CurrentCommand, len(commands))
+}
+
+// SetCompletionTime is a utility method for setting the CompletionTime field to Now.
+func (s *RuntimeTaskStatus) SetCompletionTime() {
+	now := metav1.Now()
+	s.CompletionTime = &now
+
+	// ensure all the status attributes are clean after complete
+	s.Paused = false
```

Comment:
nit:
it feels a bit odd that the complication time setter method updates the running state (paused) of the Status object.

Reply:
As per comment, this is for ensuring a clean state in corner cases; might be we can get rid of this when implementing the controller